server/subscriptions.go

```diff
@@ -2,67 +2,158 @@ package main
 
 import (
 	"bytes"
+	"context"
 	"encoding/json"
 	"fmt"
 
-	"github.com/mattermost/mattermost-server/plugin"
+	"github.com/mattermost/mattermost-server/mlog"
+
+	"github.com/google/go-github/github"
 )
 
 const (
 	SUBSCRIPTIONS_KEY = "subscriptions"
 )
 
+type Subscription struct {
+	ChannelID string
+	Features  string
+}
+
 type Subscriptions struct {
-	Repositories map[string][]string
+	Repositories map[string][]*Subscription
+}
+
+func (p *Plugin) Subscribe(ctx context.Context, githubClient *github.Client, userId, ownerAndRepo, channelID, features string) error {
+	_, owner, repo := parseOwnerAndRepo(ownerAndRepo, p.EnterpriseBaseURL)
+
+	if owner == "" {
+		return fmt.Errorf("Invalid repository")
+	}
+
+	if err := p.checkOrg(owner); err != nil {
+		return err
+	}
+
+	if result, _, err := githubClient.Repositories.Get(ctx, owner, repo); result == nil || err != nil {
+		if err != nil {
+			mlog.Error(err.Error())
+		}
+		return fmt.Errorf("Unknown repository %s/%s", owner, repo)
+	}
+
+	sub := &Subscription{
+		ChannelID: channelID,
+		Features:  features,
+	}
+
+	if err := p.AddSubscription(fmt.Sprintf("%s/%s", owner, repo), sub); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Plugin) AddSubscription(repo string, sub *Subscription) error {
+	subs, err := p.GetSubscriptions()
+	if err != nil {
+		return err
+	}
+
+	repoSubs := subs.Repositories[repo]
+	if repoSubs == nil {
+		repoSubs = []*Subscription{sub}
+	} else {
+		exists := false
+		for index, s := range repoSubs {
+			if s.ChannelID == sub.ChannelID {
+				repoSubs[index] = sub
+				exists = true
+				break
+			}
+		}
+
+		if !exists {
+			repoSubs = append(repoSubs, sub)
+		}
+	}
+
+	subs.Repositories[repo] = repoSubs
+
+	err = p.StoreSubscriptions(subs)
+	if err != nil {
+		return err
+	}
+
+	return nil
 }
 
-func NewSubscriptionsFromKVStore(store plugin.KeyValueStore) (*Subscriptions, error) {
+func (p *Plugin) GetSubscriptions() (*Subscriptions, error) {
 	var subscriptions *Subscriptions
 
-	value, err := store.Get(SUBSCRIPTIONS_KEY)
+	value, err := p.API.KVGet(SUBSCRIPTIONS_KEY)
 	if err != nil {
 		return nil, err
 	}
 
 	if value == nil {
-		subscriptions = &Subscriptions{}
+		subscriptions = &Subscriptions{Repositories: map[string][]*Subscription{}}
 	} else {
 		json.NewDecoder(bytes.NewReader(value)).Decode(&subscriptions)
 	}
 
 	return subscriptions, nil
 }
 
-func (s *Subscriptions) StoreInKVStore(store plugin.KeyValueStore) error {
+func (p *Plugin) StoreSubscriptions(s *Subscriptions) error {
```

Comment:
Is there a potential for a race condition here? Either between multiple requests to the plugins, or when the plugin is deployed in an HA environment?

Reply:
Yes there is, and also in other places such as updating a user's settings (though that one's less likely to be an issue). We could probably split the subscriptions by repo to reduce the chance of a race causing problems but that won't eliminate it. To eliminate it we'd need a way of atomically updating them in the KV store. Thoughts?

Reply:
I'm not sure of a way around this for the KV store. It seems like it might be easier just to formally give plugins database access and use transactions.

Reply:
Yeah that's an option, something to look into further for sure